rlberry/agents/torch/dqn/dqn_utils.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import logging
+from rlberry.utils.jit_setup import numba_jit
+
+
+logger = logging.getLogger(__name__)
+
+
+def polynomial_schedule(
+    init_value: float,
+    end_value: float,
+    power: float,
+    transition_steps: int,
+    transition_begin: int = 0,
+):
+    """Constructs a schedule with polynomial transition from init to end value.
+
+    Notes
+    -----
+
+    Function taken from: https://github.com/deepmind/optax/blob/master/optax/_src/schedule.py,
+    which is licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+
+    Modifications with respect to source:
+
+    * Remove chex typing from the arguments.
+    * Log using a logger :code:`logging.getLogger(__name__)` instead of :code:`logging.info()`.
+    * Changed documentation style.
+
+    Parameters
+    ----------
+    init_value: float
+        Initial value for the scalar to be annealed.
+    end_value: float
+        End value of the scalar to be annealed.
+    power: float
+        The power of the polynomial used to transition from init to end.
+    transition_steps: float
+        Number of steps over which annealing takes place,
+        the scalar starts changing at `transition_begin` steps and completes
+        the transition by `transition_begin + transition_steps` steps.
+        If `transition_steps <= 0`, then the entire annealing process is disabled
+        and the value is held fixed at `init_value`.
+    transition_begin: float
+        Must be positive. After how many steps to start annealing
+        (before this many steps the scalar value is held fixed at `init_value`).
+
+    Returns
+    -------
+    schedule: Callable[[int], float]
+        A function that maps step counts to values.
+    """
+    if transition_steps <= 0:
+        logger.info(
+            "A polynomial schedule was set with a non-positive `transition_steps` "
+            "value; this results in a constant schedule with value `init_value`."
+        )
+        return lambda count: init_value
+
+    if transition_begin < 0:
+        logger.info(
+            "An exponential schedule was set with a negative `transition_begin` "
+            "value; this will result in `transition_begin` falling back to `0`."
+        )
+        transition_begin = 0
+
+    def schedule(count):
+        count = np.clip(count - transition_begin, 0, transition_steps)
+        frac = 1 - count / transition_steps
+        return (init_value - end_value) * (frac**power) + end_value
+
+    return schedule
+
+
+@numba_jit
```

Comment:
Could you share/link the code using `scipy.signal.lfilter`?  It might be a good idea to use `scipy.signal.lfilter` here, but I think the advantage of numba + loop is that it can be more readable (e.g. seeing clearly what happens when lambda_ = 1 or 0) and there should be no loss of performance.

Reply:
You're right. The `signal.lfilter` approach is famously opaque to the user - maybe it is worth it to support `numba` so we can have more readable and flexible code.

On another note, why do we use `numba` for JIT compilation? Is it a requirement for the base version of `rlberry`?

Reply:
Yes, numba is included in the default installation mode. If numba is not installed, the code still runs without JIT compilation (I think readthedocs was not working with numba at some point, so we made it optional, that's why we don't import numba directly).

Numba is nice because it only requires us to add a decorator to a function, and we don't need to use another interpreter (as PyPy), or to change the code (as cython).